### ISSUE TITLE
Feature Idea: reduce tokens by overwriting old ones

### ISSUE THREAD

**Issue by Moosems** [Moosems/salve#35]
Feature report:

Instead of just putting overwriting tokens at the end of the list when highlighting actually update the tokens and push in the new tokens with a overwrite_old_tokens_with_new function that takes the list of old tokens and new ones to add and overwrites them precisely. This should also allow some tokens to coexist like hidden chars and string token in the same ares

**Comment by Moosems** [comment]
Specifically links and hidden chars

**Comment by Moosems** [comment]
Can be partially completed (one pr to implement the method and another to use it (Waiting on PR #34 to be pushed to v0.6.0 branch to add the docstring tokens with the method))

**Comment by Moosems** [comment]
```python
existing_tokens = [
    ((1, 0), 1, "String"),
    ((1, 1), 6, "String"),
    ((1, 7), 1, "String"),
    ((1, 8), 3, "String"),
    ((1, 11), 1, "String"),
    ((1, 12), 1, "String"),
    ((2, 0), 3, "String"),
    ((3, 0), 4, "Name"),
    ((3, 4), 1, "Punctuation"),
    ((3, 5), 4, "Name"),
    ((3, 9), 1, "Punctuation"),
    ((4, 0), 3, "String"),
    ((5, 0), 3, "Keyword"),
    ((5, 4), 1, "Name"),
    ((5, 5), 1, "Punctuation"),
    ((5, 6), 1, "Punctuation"),
    ((5, 7), 1, "Punctuation"),
    ((5, 9), 4, "Keyword"),
]

new_tokens = [
    ((2, 0), 3, "String"),
    ((3, 0), 10, "String"),
    ((4, 0), 3, "String"),
]

output = overwrite_tokens(old_tokens=existing_tokens, new_tokens=new_tokens)

assert output == [
    ((1, 0), 1, "String"),
    ((1, 1), 6, "String"),
    ((1, 7), 1, "String"),
    ((1, 8), 3, "String"),
    ((1, 11), 1, "String"),
    ((1, 12), 1, "String"),
    ((2, 0), 3, "String"),
    ((3, 0), 10, "String"),
    ((4, 0), 3, "String"),
    ((5, 0), 3, "Keyword"),
    ((5, 4), 1, "Name"),
    ((5, 5), 1, "Punctuation"),
    ((5, 6), 1, "Punctuation"),
    ((5, 7), 1, "Punctuation"),
    ((5, 9), 4, "Keyword"),
]
```

**Comment by Moosems** [comment]
Hidden chars never call this function, they are truly just added to the end, the docstring/comments are used though

**Comment by Moosems** [comment]
My working example (tests are not 100% conclusive):
```python
Token = tuple[tuple[int, int], int, str]

_RangeTokenDict = dict[int, list[tuple[range, str]]]


def merge_tokens(tokens: list[Token]) -> list[Token]:
    output_tokens: list[Token] = []
    depth: int = 0
    for token in tokens:
        # Deal with basic edge case
        if depth == 0:
            output_tokens.append(token)
            depth += 1
            continue

        previous_token = output_tokens[-1]

        # Get our boolean checks
        same_token_type: bool = previous_token[2] == token[2]
        same_line: bool = previous_token[0][0] == token[0][0]
        neighboring_tokens: bool = (
            previous_token[0][1] + previous_token[1] == token[0][1]
        )

        # Determine if tokens should be merged
        if not (same_token_type and same_line and neighboring_tokens):
            output_tokens.append(token)
            depth += 1
            continue

        # Replace previous token with new token (we don't increase depth because we are substituting, not adding)
        new_token: Token = (
            (token[0][0], previous_token[0][1]),
            previous_token[1] + token[1],
            token[2],
        )
        output_tokens[-1] = new_token
    return output_tokens

def overwrite_tokens(old_tokens: list[Token], new_tokens: list[Token]):
    output_tokens: list[Token] = []
    dont_add_tokens: list[Token] = []
    for new_token in new_tokens:
        for old_token in old_tokens:
            same_token: bool = old_token == new_token
            if same_token:
                continue

            same_line: bool = old_token[0][0] == new_token[0][0]
            can_add_token: bool = not old_token in dont_add_tokens
            if not same_line:
                if can_add_token:
                    output_tokens.append(old_token)
                continue

            # Check if the ranges overlap and if so either (remove the old_token and add to don't add list) or,
            # if part of the token is out of the new_token_range, remove the part in the new tokens range

            old_token_end: int = old_token[0][1] + old_token[1]
            new_token_end: int = new_token[0][1] + new_token[1]

            front_overlaps: bool = new_token[0][1] <= old_token_end and not old_token_end > new_token_end
            end_overlaps: bool = new_token_end >= old_token[0][1]
            full_overlap: bool = front_overlaps and end_overlaps

            if not (front_overlaps or end_overlaps):
                continue

            dont_add_tokens.append(old_token)

            while old_token in output_tokens:
                output_tokens.remove(old_token)

            if full_overlap:
                continue

            # If we are here if means its a partial overlap
            if front_overlaps:
                created_token: Token = (
                    (new_token[0][0], old_token[0][1]), new_token[0][1] - old_token[1], old_token[2]
                )
                output_tokens.append(created_token)
                dont_add_tokens.append(created_token)
                continue

            created_token: Token = (
                (new_token[0][0], new_token_end), old_token_end - new_token_end, old_token[2]
            )
            output_tokens.append(created_token)
            dont_add_tokens.append(created_token)

        output_tokens.append(new_token)

    output_tokens = sorted(set(output_tokens))
    return output_tokens


def overwrite_and_merge_tokens(
    old_tokens: list[Token], new_tokens: list[Token]
) -> list[Token]:
    merged_old_tokens: list[Token] = merge_tokens(old_tokens)
    merged_new_tokens: list[Token] = merge_tokens(new_tokens)
    output_tokens: list[Token] = overwrite_tokens(
        merged_old_tokens, merged_new_tokens
    )

    output_tokens = merge_tokens(output_tokens)
    return output_tokens



existing_tokens: list[Token] = [
    ((2, 0), 5, "Name"),
    ((2, 5), 1, "Punctuation"),
    ((2, 6), 1, "String"),
    ((2, 7), 3, "String"),
    ((2, 10), 1, "String"),
    ((2, 11), 1, "Punctuation"),
    ((3, 0), 3, "String"),
    ((4, 0), 4, "Name"),
    ((4, 4), 1, "Punctuation"),
    ((4, 5), 4, "Name"),
    ((4, 9), 1, "Punctuation"),
    ((4, 9), 5, "Test"),
    ((4, 11), 5, "Test"),
    ((5, 0), 3, "String"),
    ((6, 0), 3, "Keyword"),
    ((6, 4), 1, "Name"),
    ((6, 5), 1, "Punctuation"),
    ((6, 6), 1, "Punctuation"),
    ((6, 7), 1, "Punctuation"),
    ((6, 9), 4, "Keyword"),
]

new_tokens: list[Token] = [
    ((3, 0), 3, "String"),
    ((4, 0), 10, "String"),
    ((5, 0), 3, "String"),
]

output: list[Token] = overwrite_and_merge_tokens(existing_tokens, new_tokens)

assert output == [
    ((2, 0), 5, "Name"),
    ((2, 5), 1, "Punctuation"),
    ((2, 6), 5, "String"),
    ((2, 11), 1, "Punctuation"),
    ((3, 0), 3, "String"),
    ((4, 0), 10, "String"),
    ((4, 10), 4, "Test"),  # Super important
    ((4, 11), 5, "Test"),  # Super important
    ((5, 0), 3, "String"),
    ((6, 0), 3, "Keyword"),
    ((6, 4), 1, "Name"),
    ((6, 5), 3, "Punctuation"),
    ((6, 9), 4, "Keyword"),
]
``

**Comment by Moosems** [comment]
Part one is now up on branch `implement-token-overwriting`. Will push to 0.7.0 branch at 3 pm mst tomorrow after pulling main (which will have merged 0.6.0 and #34) to 0.7.0 with part two of the plan implemented after pulls. 0.8.0 should also feature a major refactor of the highlight code to simplify it and make it easier to maintain. Will make a separate issue for this.
